src/ConstructTgtTube_tool.cc

```diff
@@ -50,19 +62,112 @@ namespace mu2e {
 
     // Extract box information from the config file.
 
-    G4bool boxVisible        = _config.getBool("box.visible",true);
-    G4bool boxSolid          = _config.getBool("box.solid",true);
+     G4bool boxVisible        = _config.getBool("box.visible",true);
+     G4bool boxSolid          = _config.getBool("box.solid",true);
 
+     // Due to creating multiple tubes or boxes,
+     // Using material finder may not be neccesary as you can directly fin the material Using findaMaterialOrThrow.
+    MaterialFinder materialFinder(_config);
+    // arrays of information for each cylinder are declared
+    // information from the geometry file(ProdTgt4) tube is retrieved using the online function
+    G4bool tubeVisible        = _config.getBool("tube.visible",true);
+    G4bool tubeSolid          = _config.getBool("tube.solid",true);
+    string tubeBase = "tube";
+    G4Colour  orange  (.75, .55, .0);
+    int numberOfTubes =  _config.getInt("tube.numberOfTubes");
+    vector <double> tubeRIns;
+    _config.getVectorDouble("tube.rIn",tubeRIns);
+    vector <double> tuberOuts;
+    _config.getVectorDouble("tube.rOut", tuberOuts);
+    vector <double> halfLengths;
+    _config.getVectorDouble("tube.halfLength", halfLengths);
+    vector <double> phiZeroes;
+    _config.getVectorDouble("tube.phi0", phiZeroes);
+    vector <double> phiSpans;
+    _config.getVectorDouble("tube.phiSpan",phiSpans );
+    vector <double> phiRotZs;
+    _config.getVectorDouble("tube.phiRotZ",phiRotZs);
+    vector <double> signs;
+    _config.getVectorDouble("tube.sign",signs);
+    vector <int> copyNumbers;
+    _config.getVectorInt("tube.copyNumber", copyNumbers);
+    //varaibales to be used to retrieve the variables from the vector string, convert them to doubles
+    vector<string> centerCoordinates;
+    _config.getVectorString("tube.centerInWorld", centerCoordinates);
+    vector <string> material;
+    _config.getVectorString("tube.wallMaterialName" ,material);
+    // traverse the string of center coordinates to convert them to vector arrays with each array containing 3 doubles.
+    // get the data in a string and convert it into a vector
+   vector <array<double,3>> center;
+   string toAdd;
+   int index = 0;
+   array<double,3> currentTriplet; 
+   for (const auto& coordinate : centerCoordinates) {
+        for (char c : coordinate) {
+            if ((c >= '0' && c <= '9') || c == '.' || c == '-' || c == 'e' || c == 'E') {
```

Comment:
Looks like you're trying to cast a string as a number and include compatibility with scientific notation. Wouldn't it make more sense to use stringstream?

Try this in a ROOT prompt for an example:
```
root [0] stringstream ss("3e-3")
(std::stringstream &) @0x7f067139d000
root [1] double d = 0
(double) 0.0000000
root [2] ss >> d
(std::basic_istream<char, std::char_traits<char> >::__istream_type &) @0x7f067139d000
root [3] d
(double) 0.0030000000
```

Reply:
string stream can work. ut my function still works. What would be the benefit to using stringstream as opposed to what I have?

Reply:
It as always better to use standard libraries than hand-written code wherever possible, even if it "works" in the successful case. For example, consider if a coordinate has a typo in it: "3d-4" or "3ee-4". This could be caught with simple try/catch with stringstream, because it already comes with a method to check if the casting worked:

```
root [25] stringstream ss("3Ee-3")
(std::stringstream &) @0x7f06711b1000
root [26] d
(double) 0.0000000
root [27] ss >> d
(std::basic_istream<char, std::char_traits<char> >::__istream_type &) @0x7f06711b1000
root [28] ss.fail()
(bool) true
root [29] d = 0
(double) 0.0000000
root [30] stringstream ss("3E-3")
(std::stringstream &) @0x7f0671199000
root [31] ss >> d
(std::basic_istream<char, std::char_traits<char> >::__istream_type &) @0x7f0671199000
root [32] d
(double) 0.0030000000
root [33] ss.fail()
(bool) false
```

Reply:
I have not taken those exceptions into consideration yet. Would using string stream make it easier to check the casting and writing the try catch exception?

Reply:
Yes, definitely. Take a look at the ROOT example above (ROOT can be used as a REPL for trying out C++)

Reply:
I will work on those changes for the method.

Reply:
I have pushed the code with the string stream. You should be able to see my commit.

Reply:
Very good, thank you! Can you clean up the comments, please? Don't worry about saving code that won't be used anymore, we can always go back to the commit if we need it

Reply:
When it comes to the cleanup, you are okay with me deleting the code that constructs the boxes because it is in your original version of your repository? I will clean up my comments. However, I do not know if I should keep the comments at the top regarding the changes I made to the code.

Reply:
Ah good question, thanks for checking. Only worry about deleting comment blocks of unused code. Yes, please feel free to delete the code on constructing boxes. Please keep written comments, as long as they are still relevant to the code.

Reply:
Should I delete the remarked code that creates the mesh without Union Solid. I would consider keeping it because it could be used to experiment with more combinations with the wire mesh. On the other hand, my previous commits have the code so would it not hurt to removed the remarked code not involving G4UnionSolid?

Reply:
Good point. Feel free to leave those in, as that would be useful to keep. Just add a written comment describing what it is and label it WIP (work in progress) or something similar

Reply:
I can add that remark if the reamrks that I put on there do not seem to rpovide the best explanation.

Reply:
I pushed my code with the unecesarry remarks removed. See if there are any other areas I should try to remark that I may need or have forgotten.

Reply:
Very good, thank you!